### PR TITLE
Support EnumerableShardingSpec in ShardedTensor.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59776 Add RRef support to ShardedTensor.
* **#59061 Support EnumerableShardingSpec in ShardedTensor.**

Overall Design: https://github.com/pytorch/pytorch/issues/55207

This PR builds upon https://github.com/pytorch/pytorch/pull/58517 and
https://github.com/pytorch/pytorch/pull/57409 to support creating a
ShardedTensor using EnumerableShardingSpec.

Differential Revision: [D28734551](https://our.internmc.facebook.com/intern/diff/D28734551/)

### CODE REVIEW

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 1]
```diff
@@ -283,3 +289,269 @@ def test_insufficient_sharding_dims(self):
...
+            sharding_metadata = sharded_tensor.sharding_metadata()
+            self.assertEqual(2, len(sharding_metadata))
+            for rank, shard_metadata in enumerate(sharding_metadata):
+                self.assertEqual((rank * 5, 0), shard_metadata.shard_offsets)
+                self.assertEqual((5, 5), shard_metadata.shard_lengths)
+                self.assertEqual(f'rank:{rank * 2}/cuda:{rank * 2 + 1}', shard_metadata.placement)
+
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_multiple_local_shards(self):
```
Comment (SciPioneer) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:507]: Can you add a few more test cases to answer the following questions?
1) Do we allow overlap among different shards?
2) Do we allow the union of all the shards smaller than the input large tensor?

Comment (pritamdamania87) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:507]: I don't think we support any overlaps amongst any shards at the moment. All the shards should be mutually exclusive.

Comment (SciPioneer) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:507]: Can we add a note at somewhere to declare this or shall we add a check to fail such invalid input?

Comment (pritamdamania87) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:507]: Makes sense I will update the docs to add this, regarding the check we have the check and a test for this as well: https://github.com/pytorch/pytorch/blob/master/test/distributed/_sharding_spec/test_sharding_spec.py#L158.

**Review by wayi1 — APPROVED** [review]

File: torch/distributed/_sharding_spec/api.py [hunk 2]
```diff
@@ -208,12 +208,12 @@ def _check_shard_pair_overlap(shard1: ShardMetadata, shard2: ShardMetadata):
 
         return True
 
-    def check_tensor(self, tensor: torch.Tensor) -> None:
+    def check_tensor(self, tensor_dims) -> None:
```
Comment (cbalioglu) [inline torch/distributed/_sharding_spec/api.py:211]: Nit: `tensor_dims` can be type annotated as `tensor_dims: Sequence[int]`.

**Pull request closed without merging**